From c4e1f54c4e6c121e00555cc289bda2395151c354 Mon Sep 17 00:00:00 2001
From: Bjorn Jonsson <bjornjon@gmail.com>
Date: Sun, 13 Mar 2016 10:18:24 +0000
Subject: [PATCH 199/517] [SPARK-13810][CORE] Add Port Configuration
 Suggestions on Bind Exceptions

## What changes were proposed in this pull request?
Currently, when a java.net.BindException is thrown, it displays the following message:

java.net.BindException: Address already in use: Service '$serviceName' failed after 16 retries!

This change adds port configuration suggestions to the BindException, for example, for the UI, it now displays

java.net.BindException: Address already in use: Service 'SparkUI' failed after 16 retries! Consider explicitly setting the appropriate port for 'SparkUI' (for example spark.ui.port for SparkUI) to an available port or increasing spark.port.maxRetries.

## How was this patch tested?
Manual tests

Author: Bjorn Jonsson <bjornjon@gmail.com>

Closes #11644 from bjornjon/master.

(cherry picked from commit 515e4afbc7ec957609451ea75772d6ef1b914908)
Signed-off-by: Sean Owen <sowen@cloudera.com>
(cherry picked from commit 5e08db30d55a6be8996d9de452720f9a465ebcc5)
---
 .../main/scala/org/apache/spark/util/Utils.scala   |    6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/core/src/main/scala/org/apache/spark/util/Utils.scala b/core/src/main/scala/org/apache/spark/util/Utils.scala
index a8d515a..842a77c 100644
--- a/core/src/main/scala/org/apache/spark/util/Utils.scala
+++ b/core/src/main/scala/org/apache/spark/util/Utils.scala
@@ -1992,8 +1992,10 @@ private[spark] object Utils extends Logging {
       } catch {
         case e: Exception if isBindCollision(e) =>
           if (offset >= maxRetries) {
-            val exceptionMessage =
-              s"${e.getMessage}: Service$serviceString failed after $maxRetries retries!"
+            val exceptionMessage = s"${e.getMessage}: Service$serviceString failed after " +
+              s"$maxRetries retries! Consider explicitly setting the appropriate port for the " +
+              s"service$serviceString (for example spark.ui.port for SparkUI) to an available " +
+              "port or increasing spark.port.maxRetries."
             val exception = new BindException(exceptionMessage)
             // restore original stack trace
             exception.setStackTrace(e.getStackTrace)
-- 
1.7.9.5

